tensorflow_datasets/translate/opus.py

```diff
@@ -0,0 +1,255 @@
+    """
+    name = kwargs.get("name", "%s-%s for %s" % (
+        language_pair[0], language_pair[1], ', '.join(subsets)))
+
+    description = name + " documents"
+
+    super(OpusConfig, self).__init__(
+        description=description, **dict(kwargs, name=name))
+
+    self.language_pair = language_pair
+    self.subsets = subsets
+
+class Opus(tfds.core.GeneratorBasedBuilder):
+  """OPUS is a collection of translated texts from the web."""
+
+  """
+  The following configurations reproduce the evaluation tasks from "Six Challenges for Neural Machine Translation" by Philipp Koehn and Rebecca Knowles (2017) https://www.aclweb.org/anthology/W17-3204.pdf"""
+
+  _KK_SUBSETS = [
+      ("medical", ["EMEA"]),
+      ("law", ["JRC-Acquis"]),
+      ("koran", ["Tanzil"]),
+      ("IT", ["GNOME", "KDE4", "PHP", "Ubuntu", "OpenOffice"]),
+      ("subtitles", ["OpenSubtitles"])
+  ]
+
+  BUILDER_CONFIGS = [
+      OpusConfig(
+          version=tfds.core.Version('0.1.0'),
+          language_pair=("de", "en"),
+          subsets=subsets,
+          name=name
+      ) for name, subsets in _KK_SUBSETS
+  ]
+
+  @property
+  def subsets(self):
+    # Return only the datasets that exist for the language pair.
+    source, target = self.builder_config.language_pair
+    filtered_subsets = []
+    for dataset in [DATASET_MAP[name] for name in self.builder_config.subsets]:
+      if (source, target) in dataset.language_pairs:
+        filtered_subsets.append(dataset)
+
+    return filtered_subsets
+
+  def _info(self):
+    src, target = self.builder_config.language_pair
+    return tfds.core.DatasetInfo(
+        builder=self,
+        description=_DESCRIPTION + "\n" + self.builder_config.description,
+        features=tfds.features.Translation(
+            languages=self.builder_config.language_pair),
+        supervised_keys=(src, target),
+        homepage='http://opus.nlpl.eu/',
+        citation=_CITATION,
+    )
+
+  def _split_generators(self, dl_manager):
+    source, target = self.builder_config.language_pair
+    file_ext = "%s-%s"%(source, target)
+
+    subsets = []
+    for item in self.subsets:
+      dl_dir = dl_manager.download_and_extract(
+          os.path.join(item.url, "%s.txt.zip"%file_ext))
+      source_file = os.path.join(
+          dl_dir, "%s.%s.%s"%(item.name, file_ext, source))
+      target_file = os.path.join(
+          dl_dir, "%s.%s.%s"%(item.name, file_ext, target))
+      subsets.append({
+          "name": item.name,
+          "source_file": source_file,
+          "target_file": target_file
+      })
+
+    return [
+        tfds.core.SplitGenerator(
+            name=tfds.Split.TRAIN,
+            gen_kwargs={"subsets": subsets}
+        )
+    ]
+
+  def _generate_examples(self, subsets):
+    for item in subsets:
+      logging.info("Generating examples from: %s", item["name"])
+      source_file = item["source_file"]
+      target_file = item["target_file"]
+
+      with tf.io.gfile.GFile(source_file) as f:
```

Comment:
Are these small enough that we won't run into memory issues? Would it be better to iterate one line at a time?

Reply:
Done.